### ISSUE TITLE
[Suggestion] Add support for printing Graphviz DOT

### ISSUE THREAD

**Issue by sideeffffect** [kitlangton/zio-magic#30]
It could be worthwhile to support another way to "visualize" the graph: print Graphviz DOT description.
https://graphviz.org/doc/info/lang.html
https://graphviz.org/pdf/dotguide.pdf

Once the DOT description is available, people can turn them into image.

Having a nice image could be useful for having and overview of the modules as they are, or if debugging an issue of missing module(s).

And there are also Java libraries that could visualize them, if we decide to do that in `zio-magic` -- I don't know, it could be neat :wink: 
https://github.com/nidi3/graphviz-java
https://github.com/eclipse/elk

**Comment by kitlangton** [comment]
Oh! That's an awesome idea :)

**Comment by alphaho** [comment]
My two cents in this topic, there's a more light-weight alternative that I usually used to generate diagrams in my previous projects: https://github.com/mermaid-js/mermaid

It just requires a `<script>` tag to import the mermaid-js and the diagram's text description. This makes it quite easy to embed into different places.  And IMHO, the [text syntax](https://mermaid-js.github.io/mermaid/#/n00b-syntaxReference?id=diagram-syntax) is pretty easy to understand comparing to the DOT language.
A quick demo using the README.md's example:

[![](https://mermaid.ink/img/eyJjb2RlIjoiZ3JhcGggVERcbiAgICBBW0Nha2UuTGl2ZV0gLS0-IEJbQ2hvY29sYXRlLmxpdmVdXG4gICAgQSAtLT4gQ1tDaG9jb2xhdGUubGl2ZV1cbiAgICBCIC0tPiBEW1Nwb29uLmxpdmVdXG4gICAgQiAtLT4gRVtDb25zb2xlLmxpdmVdXG4gICAgQyAtLT4gRltCbG9ja2luZy5saXZlXVxuICAgICIsIm1lcm1haWQiOnsidGhlbWUiOiJkZWZhdWx0In0sInVwZGF0ZUVkaXRvciI6ZmFsc2V9)](https://mermaid-js.github.io/mermaid-live-editor/#/edit/eyJjb2RlIjoiZ3JhcGggVERcbiAgICBBW0Nha2UuTGl2ZV0gLS0-IEJbQ2hvY29sYXRlLmxpdmVdXG4gICAgQSAtLT4gQ1tDaG9jb2xhdGUubGl2ZV1cbiAgICBCIC0tPiBEW1Nwb29uLmxpdmVdXG4gICAgQiAtLT4gRVtDb25zb2xlLmxpdmVdXG4gICAgQyAtLT4gRltCbG9ja2luZy5saXZlXVxuICAgICIsIm1lcm1haWQiOnsidGhlbWUiOiJkZWZhdWx0In0sInVwZGF0ZUVkaXRvciI6ZmFsc2V9)
(click the image to go to the live editor)

**Comment by sideeffffect** [comment]
@alphaho That's a very good idea! With mermaid, `zio-magic` could only spit out a small html file with the graph description and link to the JavaScript library which would do the rest of the job.

Btw, we're using Mermaid in ZIO Prelude:
https://github.com/zio/zio-prelude/blob/master/docs/overview/diagrams.md
https://zio.github.io/zio-prelude/docs/overview/overview_diagrams

**Comment by sideeffffect** [comment]
@alphaho That's a very good idea! With mermaid, `zio-magic` could only spit out a small html file with the graph description and a link to the JavaScript library which would do the rest of the job.

Btw, we're using Mermaid in ZIO Prelude:
https://github.com/zio/zio-prelude/blob/master/docs/overview/diagrams.md
https://zio.github.io/zio-prelude/docs/overview/overview_diagrams

**Comment by alphaho** [comment]
Or we may even only generate a link to the mermaid online editor by Base64 encoding the payload like: `https://mermaid.ink/img/<Base64-encoded-payload>` 

The payload from the above image is like this:

```json
{"code":"graph TD\n    A[Cake.Live] --> B[Chocolate.live]\n    A --> C[Chocolate.live]\n    B --> D[Spoon.live]\n    B --> E[Console.live]\n    C --> F[Spoon.live]\n    F --> G[Blocking.live]\n    D --> H[Blocking.live]\n    ","mermaid":{"theme":"default"},"updateEditor":false}
```

**Comment by kitlangton** [comment]
I implemented this. I just haven't exposed the generated mermaid link in the debug output. 
I need to figure out a good way to optionally show/hide it / some means of configuring the debug macro. Open to suggestions for an API for this :)

**Comment by sideeffffect** [comment]
Splendid work, Kit!
But I would be worried that (at least in some circumstances) the generated link would be too big, which would make it impractical to give it to users.
What do you think about generating a small HTML file to `target/` and printing a link to that instead?

**Comment by kitlangton** [comment]
@sideeffffect https://github.com/zio/zio/pull/5449
🥳

**Comment by kitlangton** [comment]
Closing, due to having implemented in ZIO 2.0
